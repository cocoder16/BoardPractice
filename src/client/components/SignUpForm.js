import React from 'react';

const SignUpForm = ({
    spanMessage, onInputChange, onOverlapCheck, onFormSubmit, isModify, 
    onDeleteMode, isDeleteMode, onDelete, deleteFailedMessage
}) => {
    let idDisabled, idOverlap, submitValue, messageId;
    if (isModify) {
        idDisabled = true;
        idOverlap = null;
        messageId = <span></span>
        submitValue = 'save'
    } else {
        idDisabled = false;
        idOverlap = <button type='button' name='idCheck' onClick={onOverlapCheck}>Check Availability</button>;
        messageId = <span>{spanMessage.id}</span>;
        submitValue = 'Register'
    }

    const template = 
        <form onSubmit={onFormSubmit}>
            <p>ID</p>
            <input type='text' maxLength='12' name='id'
                onChange={onInputChange} disabled={idDisabled}
            />
            {idOverlap}
            <div className='fail-message'>
                {messageId}
            </div>
            <p>Password</p>
            <input type='password' maxLength='14' name='pw'
                onChange={onInputChange}
            />
            <div className='fail-message'>
                <span >{spanMessage.pw}</span>
            </div>
            <p>Confirm Password</p>
            <input type='password' maxLength='14' name='pwConfirm'
                onChange={onInputChange}
            />
            <div className='fail-message'>
                <span>{spanMessage.pwConfirm}</span>
            </div>
            <p>Nickname</p>
            <input type='text' maxLength='12' name='nickname'
                onChange={onInputChange}
            />
            <button type='button' name='nicknameCheck' onClick={onOverlapCheck}>Check Availability</button>
            <div className='fail-message'>
                <span>{spanMessage.nickname}</span>
            </div>
            <p>Email</p>
            <input type='text' maxLength='40' name='email'
                onChange={onInputChange}
            />
            <div className='fail-message'>
                <span>{spanMessage.email}</span>
            </div>
            <button type='submit' className='submit'>{submitValue}</button>
        </form>

    return (
        <div className='sign-up-form'>
            { !isModify ? <h2>Sign Up</h2> : <h2>Management</h2> }
            {template}
            { isModify &&
                <button type='button' className='remove-btn' 
                    onClick={onDeleteMode}
                >Remove Acocount</button>
            }
            { (isModify && isDeleteMode) ?
                <form onSubmit={onDelete} className='delete-form'>
                    <span>Are you sure you would like to remove this account?</span>
                    <input type='password' className='pw' maxLength='14'
                        placeholder='Input Password again.'/>
                    <span className='fail-message'>{deleteFailedMessage}</span><br/>
                    <button type='submit'>Delete</button>
                </form> :
                <div className='block'></div>
            }
        </div>
    )
};

export default SignUpForm;