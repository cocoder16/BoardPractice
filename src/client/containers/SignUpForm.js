import React, { Component } from 'react';
import {connect} from 'react-redux';
import * as signUpActions from '~c/store/signUp';
import { createUser } from '~c/services/users';

class SignUpForm extends Component {
    handleInputChange = (e) => {
        e.preventDefault();
        const { name, value } = e.target;
        this.props.inputChange({name: name, value: value});
    }

    overlapCheck = async (e) => {
        const { name } = e.target;
        await this.props.beforeOverlapCheck(name);
        if (!this.props.goOverlapCheck) return null;
        this.props.overlapCheck(name);
    }

    alertReject () {
        alert('가입 양식을 다시 확인해주세요.');
    }

    handleFormSubmit = async (e) => {
        e.preventDefault();

        await this.props.formValidationInput();
        if (!this.props.pass) {
            console.log('fail form validation due to input value');
            this.alertReject();
            return false;
        }

        await this.props.formValidationOverlap();
        if (!this.props.pass) {
            console.log('fail form validation due to overlap');
            this.alertReject();
            return false;
        }

        const { id, pw, nickname, email } = this.props;
        const formData = new FormData();
        formData.append('id', id);
        formData.append('pw', pw);
        formData.append('nickname', nickname);
        formData.append('email', email);
        await createUser(formData);
    }

    render() {
        const { span } = this.props;

        return (
            <form onSubmit={this.handleFormSubmit}>
                <table><tbody>
                    <tr>
                        <td data-testid='fh_id'>아이디</td>
                        <td>
                            <input data-testid='input_id' type='text' maxLength='12' name='id'
                                onChange={this.handleInputChange}/>
                        </td>
                        <td><input data-testid='btn_idConfirm' type='button' value='중복체크' name='idCheck'
                                onClick={this.overlapCheck}/>
                        <span data-testid='span_idCheck'>{span.id}</span></td>
                    </tr>
                    <tr>
                        <td data-testid='fh_pw'>비밀번호</td>
                        <td>
                            <input data-testid='input_pw' type='password' maxLength='14' name='pw'
                                onChange={this.handleInputChange}/>
                        </td>
                        <td><span data-testid='span_pwCheck'>{span.pw}</span></td>
                    </tr>
                    <tr>
                        <td data-testid='fh_pwConfirm'>비밀번호 확인</td>
                        <td>
                            <input data-testid='input_pwConfirm' type='password' maxLength='14' name='pwConfirm'
                                onChange={this.handleInputChange}/>
                        </td>
                        <td><span data-testid='span_pwConfirmCheck'>{span.pwConfirm}</span></td>
                    </tr>
                    <tr>
                        <td data-testid='fh_nick'>닉네임</td>
                        <td>
                            <input data-testid='input_nick' type='text' maxLength='12' name='nickname'
                                onChange={this.handleInputChange}/>
                        </td>
                        <td><input data-testid='btn_nickConfirm' type='button' value='중복체크' name='nicknameCheck'
                                onClick={this.overlapCheck}/>
                        <span data-testid='span_nickCheck'>{span.nickname}</span></td>
                    </tr>
                    <tr>
                        <td data-testid='fh_email'>이메일</td>
                        <td>
                            <input data-testid='input_email' type='text' maxLength='40' name='email'
                                onChange={this.handleInputChange}/>
                        </td>
                        <td><span data-testid='span_emailCheck'>{span.email}</span></td>
                    </tr>
                </tbody></table>
                <input data-testid='btn_submit' type='submit' value='가입'/>
            </form>
        )
    }
}

//props값으로 넣어줄 상태 정의
const mapStateToProps = (state) => ({
    id: state.signUp.id,
    pw: state.signUp.pw,
    pwConfirm: state.signUp.pwConfirm,
    nickname: state.signUp.nickname,
    email: state.signUp.email,
    span: {
        id: state.signUp.span.id,
        pw: state.signUp.span.pw,
        pwConfirm: state.signUp.span.pwConfirm,
        nickname: state.signUp.span.nickname,
        email: state.signUp.span.email
    },
    goOverlapCheck: state.signUp.goOverlapCheck,
    pass: state.signUp.pass
})

//props값으로 넣어줄 액션 함수들 정의
const mapDispatchToProps = (dispatch) => ({
    inputChange: (payload) => dispatch(signUpActions.inputChange(payload)),
    beforeOverlapCheck: (payload) => dispatch(signUpActions.beforeOverlapCheck(payload)),
    overlapCheck: (name) => dispatch(signUpActions.overlapCheck(name)),
    formValidationInput: () => dispatch(signUpActions.formValidationInput()),
    formValidationOverlap: () => dispatch(signUpActions.formValidationOverlap())
})

export default connect(mapStateToProps, mapDispatchToProps)(SignUpForm);