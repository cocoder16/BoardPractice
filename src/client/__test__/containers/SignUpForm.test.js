import React from 'react';
import { render, fireEvent} from '@testing-library/react';
import "@testing-library/react/dont-cleanup-after-each";
import "@testing-library/jest-dom/extend-expect";
import {Provider} from 'react-redux';
// import store from '~c/store/index';
import configureStore from 'redux-mock-store';
import SignUpForm from '~c/containers/SignUpForm';

const mockStore = configureStore();
const store = mockStore({
    signUp: {
        id: '',
        pw: '',
        pwConfirm: '',
        nickname: '',
        email: '',
        span: {
            id: '',
            pw: '',
            pwConfirm: '',
            nickname: '',
            email: ''
        },
        goOverlapCheck: '',
        pass: ''
    }
})

const tree = render(
    <Provider store={store}>
        <SignUpForm/>
    </Provider>
);

const { getByTestId } = tree;

const fh_id = getByTestId('fh_id');
const fh_pw = getByTestId('fh_pw');
const fh_pwConfirm = getByTestId('fh_pwConfirm');
const fh_nick = getByTestId('fh_nick');
const fh_email = getByTestId('fh_email');

const input_id = getByTestId('input_id');
const input_pw = getByTestId('input_pw');
const input_pwConfirm = getByTestId('input_pwConfirm');
const input_nick = getByTestId('input_nick');
const input_email = getByTestId('input_email');

const span_idCheck = getByTestId('span_idCheck');
const span_pwCheck = getByTestId('span_pwCheck');
const span_pwConfirmCheck = getByTestId('span_pwConfirmCheck');
const span_nickCheck = getByTestId('span_nickCheck');
const span_emailCheck = getByTestId('span_emailCheck');

const btn_idConfirm = getByTestId('btn_idConfirm');
const btn_nickConfirm = getByTestId('btn_nickConfirm');
const btn_submit = getByTestId('btn_submit');

it ("snapshot", () => { 
    expect(tree.container).toMatchSnapshot();
})

//id, pw, nick, email
describe("rendering", () => {    
    test ("form table headers", () => {
        expect(fh_id).toHaveTextContent('아이디');
        expect(fh_pw).toHaveTextContent('비밀번호');
        expect(fh_pwConfirm).toHaveTextContent('비밀번호 확인');
        expect(fh_nick).toHaveTextContent('닉네임');
        expect(fh_email).toHaveTextContent('이메일');
        expect(span_idCheck).toHaveTextContent('');
        expect(span_pwCheck).toHaveTextContent('');
        expect(span_pwConfirmCheck).toHaveTextContent('');
        expect(span_nickCheck).toHaveTextContent('');
        expect(span_emailCheck).toHaveTextContent('');
        expect(btn_idConfirm).toHaveValue('중복체크');
        expect(btn_nickConfirm).toHaveValue('중복체크');
    });
})

describe("realtime checker", () => {
    // 회원가입폼(제출전검사) - id 형식 실시간체크
    test ("id short", () => {
        //id입력 -> 이벤트핸들러가 호출되겠지?
        fireEvent.change(input_id, 'abcde');
        //그 결과 출력값은 이래야함.
        expect(span_idCheck).toHaveTextContent('아이디는 소문자 영문, 숫자를 조합하여 6 ~ 12자로 입력해주세요');
    });

    test ("id long", () => {
        fireEvent.change(input_id, 'abcde01234abc');
        expect(span_idCheck).toHaveTextContent('아이디는 소문자 영문, 숫자를 조합하여 6 ~ 12자로 입력해주세요');
    });

    test ("id special character", () => {
        fireEvent.change(input_id, 'abcd0$');
        expect(span_idCheck).toHaveTextContent('아이디는 소문자 영문, 숫자를 조합하여 6 ~ 12자로 입력해주세요');
    });

    test ("id 한글", () => {
        fireEvent.change(input_id, '한글아이디다');
        expect(span_idCheck).toHaveTextContent('아이디는 소문자 영문, 숫자를 조합하여 6 ~ 12자로 입력해주세요');
    });

    test ("id only eng", () => {
        fireEvent.change(input_id, 'abcdef');
        expect(span_idCheck).toHaveTextContent('아이디는 소문자 영문, 숫자를 조합하여 6 ~ 12자로 입력해주세요');
    });

    test ("id capital letter", () => {
        fireEvent.change(input_id, 'Abcd01');
        expect(span_idCheck).toHaveTextContent('아이디는 소문자 영문, 숫자를 조합하여 6 ~ 12자로 입력해주세요');
    });

    test ("id only num", () => {
        fireEvent.change(input_id, '0123456789012');
        expect(span_idCheck).toHaveTextContent('아이디는 소문자 영문, 숫자를 조합하여 6 ~ 12자로 입력해주세요');
    });

    test ("id correct", () => {
        fireEvent.change(input_id, 'abcd01');
        expect(span_idCheck).toHaveTextContent('');
    });

    // 회원가입폼(제출전검사) - 비밀번호 형식 실시간체크
    test ("pw short", () => {
        fireEvent.change(input_pw, 'abcd01#');
        expect(span_pwCheck).toHaveTextContent('비밀번호는 영문, 숫자, 특수문자를 조합하여 8 ~ 12자로 입력해주세요.');
    });

    test ("pw long", () => {
        fireEvent.change(input_pw, 'abcde01234!@#');
        expect(span_pwCheck).toHaveTextContent('비밀번호는 영문, 숫자, 특수문자를 조합하여 8 ~ 12자로 입력해주세요.');
    });

    test ("pw except eng", () => {
        fireEvent.change(input_pw, '01234!@#$%');
        expect(span_pwCheck).toHaveTextContent('비밀번호는 영문, 숫자, 특수문자를 조합하여 8 ~ 12자로 입력해주세요.');
    });

    test ("pw except num", () => {
        fireEvent.change(input_pw, 'abcde!@#$%');
        expect(span_pwCheck).toHaveTextContent('비밀번호는 영문, 숫자, 특수문자를 조합하여 8 ~ 12자로 입력해주세요.');
    });

    test ("pw except spe", () => {
        fireEvent.change(input_pw, '01234abcde');
        expect(span_pwCheck).toHaveTextContent('비밀번호는 영문, 숫자, 특수문자를 조합하여 8 ~ 12자로 입력해주세요.');
    });

    test ("pw correct", () => {
        fireEvent.change(input_pw, 'Abcde01234!@');
        expect(span_pwCheck).toHaveTextContent('');
    });

    // 회원가입폼(제출전검사) - 비밀번호 재확인 실시간체크
    test ("pw confirm not passed", () => {
        fireEvent.change(input_pw, 'abcde01234!@');
        fireEvent.change(input_pwConfirm, 'abcde0123!');
        expect(span_pwConfirmCheck).toHaveTextContent('비밀번호가 일치하지 않습니다.');
    });

    test ("pw confirm passed", () => {
        fireEvent.change(input_pw, 'abcde01234!@');
        fireEvent.change(input_pwConfirm, 'abcde01234!@');
        expect(span_pwConfirmCheck).toHaveTextContent('');
    });

    // 회원가입폼(제출전검사) - 닉네임형식 실시간체크
    test ("nick short", () => {
        fireEvent.change(input_nick, '한');
        expect(span_nickCheck).toHaveTextContent('닉네임은 한글(영문) 또는 숫자를 사용하여 2 ~ 12자로 입력해주세요.');
    });

    test ("nick long", () => {
        fireEvent.change(input_nick, '열세글자입니다012345');
        expect(span_nickCheck).toHaveTextContent('닉네임은 한글(영문) 또는 숫자를 사용하여 2 ~ 12자로 입력해주세요.');
    });

    test ("nick special character", () => {
        fireEvent.change(input_nick, '특수문자포함!');
        expect(span_nickCheck).toHaveTextContent('닉네임은 한글(영문) 또는 숫자를 사용하여 2 ~ 12자로 입력해주세요.');
    });

    test ("nick only 한글", () => {
        fireEvent.change(input_nick, '한글닉네임');
        expect(span_nickCheck).toHaveTextContent('');
    });

    test ("nick only eng", () => {
        fireEvent.change(input_nick, 'Englishnick');
        expect(span_nickCheck).toHaveTextContent('');
    });

    test ("nick only num", () => {
        fireEvent.change(input_nick, '0123');
        expect(span_nickCheck).toHaveTextContent('');
    });

    test ("nick combination", () => {
        fireEvent.change(input_nick, '한글Eng123');
        expect(span_nickCheck).toHaveTextContent('');
    });

    // 회원가입폼(제출전검사) - email형식 실시간체크
    test ("email except @ .", () => {
        fireEvent.change(input_email, '0123abccom');
        expect(span_emailCheck).toHaveTextContent('이메일 형식이 올바르지 않습니다.');
    });

    test ("email except @", () => {
        fireEvent.change(input_email, '0123abc.com');
        expect(span_emailCheck).toHaveTextContent('이메일 형식이 올바르지 않습니다.');
    });

    test ("email except .", () => {
        fireEvent.change(input_email, '0123@abccom');
        expect(span_emailCheck).toHaveTextContent('이메일 형식이 올바르지 않습니다.');
    });

    test ("email correct", () => {
        fireEvent.change(input_email, '0123@abc.com');
        expect(span_emailCheck).toHaveTextContent('');
    });
})

describe("btn checker", () => {
    // 회원가입폼(제출전검사) - 버튼으로 id 중복체크
    test ("id overlap", () => {
        fireEvent.change(input_id, 'admin0'); //DB에 저장된 값
        fireEvent.click(btn_idConfirm); //다끝나고오지않을까?
        // await wait(() => {
        //     getByText('이미 사용중인 아이디입니다. 다른 아이디를 사용해주세요.');
        // })
        expect(span_idCheck).toHaveTextContent('이미 사용중인 아이디입니다. 다른 아이디를 사용해주세요.');
    });

    test ("id overlap", () => {
        fireEvent.change(input_id, 'admin1'); //DB에 저장되지않은 값
        fireEvent.click(btn_idConfirm);
        expect(span_idCheck).toHaveTextContent('사용 가능한 아이디입니다.');
    });

    // 회원가입폼(제출전검사) - 버튼으로 닉네임 중복체크
    test ("nick overlap", () => {
        fireEvent.change(input_nick, '관리자'); //DB에 저장된 값
        fireEvent.click(btn_nickConfirm);
        expect(span_nickCheck).toHaveTextContent('이미 사용중인 닉네임입니다. 다른 닉네임을 사용해주세요.');
    });

    test ("nick overlap", () => {
        fireEvent.change(input_nick, '관리자1'); //DB에 저장되지않은 값
        fireEvent.click(btn_nickConfirm);
        expect(span_nickCheck).toHaveTextContent('사용 가능한 닉네임입니다.');
    });
})

describe("form validation onsubmit", () => {
    window.alert = () => {};
    const spyAlert = jest.spyOn(window, 'alert');
    // 회원가입폼(제출검사) - id 중복체크 및 6~12자 영문,숫자조합 검사
    test ("submit faild due to id", async () => {
        fireEvent.click(btn_submit)
        expect(spyAlert).toHaveBeenCalledTimes(1);
        expect(spyAlert).toHaveBeenCalledWith('가입 양식을 다시 확인해주세요.');
    });
    // 회원가입폼(제출검사) - pw 재확인일치 체크 및 8~12자 영문,숫자,특수문자 조합 검사
    test ("submit faild due to pw", () => {
        fireEvent.change(input_id, 'abcd0123');
        fireEvent.change(input_pw, '01234567');
        fireEvent.click(btn_submit);
        expect(spyAlert).toHaveBeenCalledTimes(1);
        expect(spyAlert).toHaveBeenCalledWith('가입 양식을 다시 확인해주세요.');
    });
    // 회원가입폼(제출검사) - 닉네임 중복체크 및 2~12자 한글 또는 영문 또는 숫자 검사
    test ("submit faild due to nick", () => {
        fireEvent.change(input_id, 'abcd0123');
        fireEvent.change(input_pw, '01234567');
        fireEvent.change(input_pw, '01234567');
        fireEvent.change(input_nick, ' ');
        fireEvent.click(btn_submit);
        expect(spyAlert).toHaveBeenCalledTimes(1);
        expect(spyAlert).toHaveBeenCalledWith('가입 양식을 다시 확인해주세요.');
    });
    // 회원가입폼(제출검사) - email형식체크
    test ("submit faild due to email", () => {
        fireEvent.change(input_id, 'abcd0123');
        fireEvent.change(input_pw, '01234567');
        fireEvent.change(input_pw, '01234567');
        fireEvent.change(input_nick, 'eee');
        fireEvent.change(input_email, 'eee@gggcom');
        fireEvent.click(btn_submit);
        expect(spyAlert).toHaveBeenCalledTimes(1);
        expect(spyAlert).toHaveBeenCalledWith('가입 양식을 다시 확인해주세요.');
    });

    test ("submit success", () => {
        fireEvent.change(input_id, 'abcd0123');
        fireEvent.change(input_pw, '01234567');
        fireEvent.change(input_pw, '01234567');
        fireEvent.change(input_nick, 'eee');
        fireEvent.change(input_email, 'eee@ggg.com');
        fireEvent.click(btn_submit);
        expect(spyAlert).toHaveBeenCalledTimes(1);
        expect(spyAlert).toHaveBeenCalledWith('환영합니다. 회원가입이 완료되었습니다.');
    });
})

