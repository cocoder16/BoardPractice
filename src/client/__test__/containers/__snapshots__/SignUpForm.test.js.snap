// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div>
  <form>
    <table>
      <tbody>
        <tr>
          <td
            data-testid="fh_id"
          >
            아이디
          </td>
          <td>
            <input
              data-testid="input_id"
              maxlength="12"
              name="id"
              type="text"
            />
          </td>
          <td>
            <input
              data-testid="btn_idConfirm"
              name="idCheck"
              type="button"
              value="중복체크"
            />
            <span
              data-testid="span_idCheck"
            />
          </td>
        </tr>
        <tr>
          <td
            data-testid="fh_pw"
          >
            비밀번호
          </td>
          <td>
            <input
              data-testid="input_pw"
              maxlength="14"
              name="pw"
              type="password"
            />
          </td>
          <td>
            <span
              data-testid="span_pwCheck"
            />
          </td>
        </tr>
        <tr>
          <td
            data-testid="fh_pwConfirm"
          >
            비밀번호 확인
          </td>
          <td>
            <input
              data-testid="input_pwConfirm"
              maxlength="14"
              name="pwConfirm"
              type="password"
            />
          </td>
          <td>
            <span
              data-testid="span_pwConfirmCheck"
            />
          </td>
        </tr>
        <tr>
          <td
            data-testid="fh_nick"
          >
            닉네임
          </td>
          <td>
            <input
              data-testid="input_nick"
              maxlength="12"
              name="nickname"
              type="text"
            />
          </td>
          <td>
            <input
              data-testid="btn_nickConfirm"
              name="nicknameCheck"
              type="button"
              value="중복체크"
            />
            <span
              data-testid="span_nickCheck"
            />
          </td>
        </tr>
        <tr>
          <td
            data-testid="fh_email"
          >
            이메일
          </td>
          <td>
            <input
              data-testid="input_email"
              maxlength="40"
              name="email"
              type="text"
            />
          </td>
          <td>
            <span
              data-testid="span_emailCheck"
            />
          </td>
        </tr>
      </tbody>
    </table>
    <input
      data-testid="btn_submit"
      type="submit"
      value="가입"
    />
  </form>
</div>
`;
